### ISSUE TITLE
Allow moved blocks to be specified outside of root main.tf

### ISSUE THREAD

**Issue by skogsbrus** [hashicorp/terraform#30255]
### Current Terraform Version

```
1.1.2
```

### Use-cases

In a large code base with multiple projects / modules, it becomes tedious to maintain a list of _all_ moves across the whole repository in one single file. 

Here are a few problems that I'm encountering when refactoring a module to use an S3 module instead of handling it on its own:

* `moved` blocks must be copied to all root `main.tf` files of all environments that use the same codebase.
* `main.tf` becomes hard to parse when filled with information that conceptually doesn't belong in the top level config. Time will compound this issue.
* The actual moves become hard to track when all moves for all resources & modules are listed in the same file. Time will compound this issue.

### Attempted Solutions

Consider a project like so:

```
environments
  production
    main.tf
  staging
    main.tf
modules
  new_module/main.tf
products
  foobar
    main.tf
    old_resource.tf
```
If `foobar` is refactored to use `new_module` instead of `old_resource.tf`, the `moved` blocks to accomplish that currently needs to be added to both `environments/production/main.tf` and `environments/staging/main.tf`. 

* If the `moved` blocks are added to `foobar/main.tf`, `terraform plan` shows that the moved resources will be destroyed and recreated.
* If the `moved` blocks are added to any of the environments/*/main.tf,  `terraform plan` shows that the moved resources will be moved as intended in the respective environment.

### Proposal

My proposal is that Terraform should support `moved` blocks outside of the root main.tf as well.

An example, where the above example has been refactored, could look like:

```
environments
  production
    main.tf
  staging
    main.tf
modules
  new_module/main.tf
products
  foobar
    main.tf
    history.tf
```

Where `history.tf` lists all of the `moved` blocks that affect resources in `foobar` since its creation.

Aside from that there are a few issues:

- It is not obvious that `moved` blocks outside of `main.tf` will be ignored. I've seen it mentioned in one sentence in the docs, but it isn't clear if it's meant as a requirement or an example.
- The moved blocks are silently ignored. Netiher `terraform plan` nor `terraform validate` complains about it.

### References

This is the only documentation I've found that kind of mentions this behaviour:

> In your root main.tf file, add moved configuration blocks for each of the resources you moved into modules in the previous step.

https://learn.hashicorp.com/tutorials/terraform/move-config

**Comment by skogsbrus** [comment]
Hey @jbardin thanks for looking at this so fast! Sorry, I did not mean that it needs to be defined in a file named `main.tf` in particular. What I'm trying to say is that it only seems to take effect if defined in the top-level module of a project?

I.e. this doesn't work

```
# products/foobar/main.tf
moved {
  from = module.products.foobar.some_resource
  to   = module.products.foobar.modules.new_module.some_resource
}
```
but this works:

```
# environments/production/main.tf
moved {
  from = module.products.foobar.some_resource
  to   = module.products.foobar.modules.new_module.some_resource
}
```

**Comment by jbardin** [comment]
Hi @skogsbrus,

Thanks for following up. The `moved` blocks are part of a _module_ configuration, just like `resource`, `locals`, `var`, etc., and a module is defined by the collection of `.tf` files within a single directory. The `moved` statements are also relative to the source of the module, so their location is important in that respect too. Having them exist at another location would mean that they could effect state differently if "imported" at different levels of the configuration. It could be quite confusing if using a module (which you might have no control over the contents of) could somehow change resources in any other module, or even the root module.

If I understand correctly, I think what you're proposing here is really that we allow the building of modules from multiple directories? If that's the case, the choice to require all files to exist in the same directory was a deliberate design choice, both for simplicity of implementation and simplicity for user's understanding. 

If I didn't correctly understand the use case, please let me know.

Thanks!

**Comment by skogsbrus** [comment]
@jbardin I think that makes sense. I might just be confused, still pretty new to Terraform (loving it so far though!).

 I'm not saying that `moved` blocks inside `modules/foobar2` should be able to move resources that are defined in `modules/foobar`. My proposal is that `moved` blocks should be able to be defined in the same module as the resources that are moved - not only in the top-level module that "owns" the whole project.

But it sounds like you're saying that this already _is_ supported?

If you think it would be useful, I can try setting up a minimal example to show what I mean :smile:

**Comment by skogsbrus** [comment]
Or could my mistake be that I didn't name the resources in relation to the current directory?

That is, should

```
# products/foobar/main.tf
moved {
  from = module.products.foobar.some_resource
  to   = module.products.foobar.modules.new_module.some_resource
}
```

instead be written like this?

```
# products/foobar/main.tf
moved {
  from = foobar.some_resource
  to   = foobar.modules.new_module.some_resource
}
```

**Comment by jbardin** [comment]
Yes, moves statements are valid inside any module, and intended to allow refactoring of resources owned by that module. The design was brought about to allow _modules_ to transparently refactor resources without any intervention from users of those modules. 

While you can move anything you want from the root module, for maintainability I would suggest pairing the move statements with whichever module is considered to be the owner of those resource configurations.

**Comment by skogsbrus** [comment]
Awesome! Thanks so much for following up on this @jbardin . I'll try to root out why I couldn't get that to work, probably a user error then :)

**Comment by skogsbrus** [comment]
Hmm, I wonder if I had the same issue as reported here, https://github.com/hashicorp/terraform/issues/30197. It wouldn't explain why it worked in the root module though.

**Comment by skogsbrus** [comment]
OK, user error confirmed!  :tada: 

For posterity I'll close off with a minimal example with an exact description of the problem I initially encountered and what I was doing wrong. Excuse any misnomers.


**Setup**

I essentially had a project structure like this

```
main.tf
module_1
```

where `main.tf` instantiated `module_1`.

**Refactoring**

I then wanted to refactor some resources inside `module_1` into a new module. So I got a project like this:

```
main.tf
module_1
module_2
```

where `main.tf` instantiated `module_1` and `module_1` instantiated `module_2`.

Running `terraform plan` here showed that all the moved resources would be destroyed and recreated (as expected).

**Configuring the moves**

Now, here's what I did wrong.

Inside `module_1`, I now added `moved` blocks like this:

```
moved {
  from = module.module_1.some_resource.this
  to   = module.module_2.some_resource.this
}
```

Why did I do it this way? Because those are the resource names that are shown in the output of `terraform plan` since that command is run from the top level (`main.tf1`).

**The correct way**

To specify `moved` blocks inside `module_1` that move resources from `module_1` to `module_2` they need to be named as follows:

```
moved {
  from = some_resource.this
  to   = module.module2.some_resource.this
}
```

This is because we're inside the "namespace" of `module1` and not the top level config.

**Why did it work in the top-level?**

Because the resource names outputted by `terraform plan` are scoped to the top-level. So inside the top-level `main.tf`, `module.module_1.some_resource.this` _is_ the correct way to specify the resource inside `module1`.
